alpa/util.py

```diff
@@ -1243,3 +1255,247 @@ def try_import_ray_worker(error: bool = False):
+                ' `ray.init("auto")`. You can also increase the timeout by '
+                "setting the ALPA_PLACEMENT_GROUP_TIMEOUT_S environment "
+                "variable. Current resources available: "
+                f"{ray.available_resources()}, resources requested by "
+                f"the placement group: {placement_group.bundle_specs}")
+        return placement_group
+    else:
+        return current_placement_group
+
+
+def get_bundle_idx(placement_group, device_ips):
+    """Get the bundle index for the placement group.
+
+    The placement group is a list of resource bundles.
+    Each bundle will be assigned to a device.
+
+    First, we need to find the bundle index with GPU resources.
+    Then, we can find the device IP for the bundle index.
+    Lastly, we sort bundle index according to the device IP list given.
+
+    Args:
+        placement_group (placement group): The placement group.
+        device_ips (list): The list of device IP addresses.
+
+    Returns:
+        list: The sorted bundle index list.
+    """
+    # get the device IP for the bundle index
+    bundle_ips = get_bundle2ip(placement_group)
+    bundle_specs = placement_group.bundle_specs
+
+    # filter out the bundle index with device (GPUs)
+    device_bundle_idx_list = [
+        i for i, bundle_spec in enumerate(bundle_specs)
+        if bundle_spec.get("GPU", 0) > 0
+    ]
+
+    if len(device_bundle_idx_list) < len(device_ips):
+        raise ValueError("The number of bundles with GPU resources "
+                         "is no less than the number of device IPs.")
+
+    # device IP -> bundle index
+    bundle_ip2idx = {bundle_ips[i]: i for i in device_bundle_idx_list}
+
+    # sorted bundle index according to the device IP list given
+    sorted_bundle_idx = [bundle_ip2idx[ip] for ip in device_ips]
+
+    return sorted_bundle_idx
+
+
+def get_placement_group_from_id(placement_group_id: str):
+    """Get the placement group from the unique id"""
+    return PlacementGroup(PlacementGroupID(hex_to_binary(placement_group_id)))
+
+
+def retrieve_placement_group():
+    """retrieve the placement group to support node affinity scheduling
+
+    If already inside the placement group, retrieve the current placement
+    group (case I). Then, if the placement group is detected globally in
+    alpa, retrieve the global placement group (case II). Last, we can try
+    to retrieve the placement group from the ray cluster global state (III).
+
+    """
+    # case 1:
+    # Get the current placement group which a task or actor is using
+    current_placement_group = get_current_placement_group()
+    if current_placement_group:
+        return current_placement_group
+
+    # case 2:
+    # Get the placement group created when alpa.init('ray')
+    if alpa.device_mesh.global_placement_group:
+        alpa_placement_group = alpa.device_mesh.global_placement_group
+        return alpa_placement_group
+
+    # case 3:
```

Comment:
When do we need case 3?
I guess case 1 is for tune integration and case 2 is for using alpa alone, but I don't understand case 3

Reply:
1. If we change https://github.com/alpa-projects/alpa/blob/792ffaa95957ed5f50f689a3c0d55169936064f0/alpa/pipeline_parallel/stage_construction.py#L463-L464 to something like ```whole_cluster_virtual_mesh = device_mesh.global_cluster.get_virtual_physical_mesh() and do not create a cluster again, can we avoid this case?
2. Getting things from the ray global state may be dangerous if the ray cluster is not running alpa only.